django/utils/translation/trans_real.py

```diff
@@ -474,14 +474,17 @@ def get_supported_language_variant(lang_code, strict=False):
     <https://www.djangoproject.com/weblog/2007/oct/26/security-fix/>.
     """
     if lang_code:
-        # If 'fr-ca' is not supported, try special fallback or language-only 'fr'.
+        # If 'zh-hant-tw' is not supported, try special fallback or subsequent
+        # language codes i.e. 'zh-hant' and 'zh'.
         possible_lang_codes = [lang_code]
         try:
             possible_lang_codes.extend(LANG_INFO[lang_code]['fallback'])
         except KeyError:
             pass
-        generic_lang_code = lang_code.split('-')[0]
-        possible_lang_codes.append(generic_lang_code)
+        lang_code_split = lang_code.split('-')
+        for i in range(len(lang_code_split) - 1, 0, -1):
+            possible_lang_codes.append('-'.join(lang_code_split[:i]))
```

Comment:
While longer, this avoids creating the extra list, string building and complex range calculations:

```suggestion
        i = None
        try:
            while i := lang_code.rindex('-', 0, i):
                possible_lang_codes.append(lang_code[:i])
        except ValueError:
            pass
```

I know it also uses the walrus operator, but Django 4.0 is targeting Python 3.8+, so it is available to us. It seems much more readable to me.

(If this isn't a performance critical path then `contextlib.suppress()` could be used to shave off two lines.)

Reply:
@pope1ni Thanks for the suggestion, but IMO it's less readable and it's more complex (as far as I'm aware) :thinking: :
<details>

```pycon
>>> import dis
>>> def test_1():
...     possible_lang_codes = []
...     lang_code = 'zh-hans-mo'
...     lang_code_split = lang_code.split('-')
...     for i in range(len(lang_code_split) - 1, 0, -1):
...         possible_lang_codes.append('-'.join(lang_code_split[:i]))
... 
>>> def test_2():
...     possible_lang_codes = []
...     lang_code = 'zh-hans-mo'
...     i = None
...     try:
...         while i := lang_code.rindex('-', 0, i):
...             possible_lang_codes.append(lang_code[:i])
...     except ValueError:
...         pass
... 
>>> dis.dis(test_1)
  2           0 BUILD_LIST               0
              2 STORE_FAST               0 (possible_lang_codes)

  3           4 LOAD_CONST               1 ('zh-hans-mo')
              6 STORE_FAST               1 (lang_code)

  4           8 LOAD_FAST                1 (lang_code)
             10 LOAD_METHOD              0 (split)
             12 LOAD_CONST               2 ('-')
             14 CALL_METHOD              1
             16 STORE_FAST               2 (lang_code_split)

  5          18 LOAD_GLOBAL              1 (range)
             20 LOAD_GLOBAL              2 (len)
             22 LOAD_FAST                2 (lang_code_split)
             24 CALL_FUNCTION            1
             26 LOAD_CONST               3 (1)
             28 BINARY_SUBTRACT
             30 LOAD_CONST               4 (0)
             32 LOAD_CONST               5 (-1)
             34 CALL_FUNCTION            3
             36 GET_ITER
        >>   38 FOR_ITER                28 (to 68)
             40 STORE_FAST               3 (i)

  6          42 LOAD_FAST                0 (possible_lang_codes)
             44 LOAD_METHOD              3 (append)
             46 LOAD_CONST               2 ('-')
             48 LOAD_METHOD              4 (join)
             50 LOAD_FAST                2 (lang_code_split)
             52 LOAD_CONST               0 (None)
             54 LOAD_FAST                3 (i)
             56 BUILD_SLICE              2
             58 BINARY_SUBSCR
             60 CALL_METHOD              1
             62 CALL_METHOD              1
             64 POP_TOP
             66 JUMP_ABSOLUTE           38
        >>   68 LOAD_CONST               0 (None)
             70 RETURN_VALUE
>>> dis.dis(test_2)
  2           0 BUILD_LIST               0
              2 STORE_FAST               0 (possible_lang_codes)

  3           4 LOAD_CONST               1 ('zh-hans-mo')
              6 STORE_FAST               1 (lang_code)

  4           8 LOAD_CONST               0 (None)
             10 STORE_FAST               2 (i)

  5          12 SETUP_FINALLY           42 (to 56)

  6     >>   14 LOAD_FAST                1 (lang_code)
             16 LOAD_METHOD              0 (rindex)
             18 LOAD_CONST               2 ('-')
             20 LOAD_CONST               3 (0)
             22 LOAD_FAST                2 (i)
             24 CALL_METHOD              3
             26 DUP_TOP
             28 STORE_FAST               2 (i)
             30 POP_JUMP_IF_FALSE       52

  7          32 LOAD_FAST                0 (possible_lang_codes)
             34 LOAD_METHOD              1 (append)
             36 LOAD_FAST                1 (lang_code)
             38 LOAD_CONST               0 (None)
             40 LOAD_FAST                2 (i)
             42 BUILD_SLICE              2
             44 BINARY_SUBSCR
             46 CALL_METHOD              1
             48 POP_TOP
             50 JUMP_ABSOLUTE           14
        >>   52 POP_BLOCK
             54 JUMP_FORWARD            20 (to 76)

  8     >>   56 DUP_TOP
             58 LOAD_GLOBAL              2 (ValueError)
             60 COMPARE_OP              10 (exception match)
             62 POP_JUMP_IF_FALSE       74
             64 POP_TOP
             66 POP_TOP
             68 POP_TOP

  9          70 POP_EXCEPT
             72 JUMP_FORWARD             2 (to 76)
        >>   74 END_FINALLY
        >>   76 LOAD_CONST               0 (None)
             78 RETURN_VALUE
```

</details>

Reply:
@felixxm Silly me... I forgot that `str.rfind()` exists which returns `-1` when not found whereas `str.rindex()` raises `ValueError`. That is actually what I wanted all along.

```suggestion
        i = None
        while (i := lang_code.rfind('-', 0, i)) >= 0:
            possible_lang_codes.append(lang_code[:i])
```

The exception handling added lots of overhead. This is much better than both the versions above if we're going to consider `dis.dis()`...
<details>

```pycon
>>> import dis
>>> def test_3():
...     possible_lang_codes = []
...     lang_code = 'zh-hans-mo'
...     i = None
...     while (i := lang_code.rfind('-', 0, i)) >= 0:
...         possible_lang_codes.append(lang_code[:i])
... 
>>> dis.dis(test_3)
  2           0 BUILD_LIST               0
              2 STORE_FAST               0 (possible_lang_codes)

  3           4 LOAD_CONST               1 ('zh-hans-mo')
              6 STORE_FAST               1 (lang_code)

  4           8 LOAD_CONST               0 (None)
             10 STORE_FAST               2 (i)

  5     >>   12 LOAD_FAST                1 (lang_code)
             14 LOAD_METHOD              0 (rfind)
             16 LOAD_CONST               2 ('-')
             18 LOAD_CONST               3 (0)
             20 LOAD_FAST                2 (i)
             22 CALL_METHOD              3
             24 DUP_TOP
             26 STORE_FAST               2 (i)
             28 LOAD_CONST               3 (0)
             30 COMPARE_OP               5 (>=)
             32 POP_JUMP_IF_FALSE       54

  6          34 LOAD_FAST                0 (possible_lang_codes)
             36 LOAD_METHOD              1 (append)
             38 LOAD_FAST                1 (lang_code)
             40 LOAD_CONST               0 (None)
             42 LOAD_FAST                2 (i)
             44 BUILD_SLICE              2
             46 BINARY_SUBSCR
             48 CALL_METHOD              1
             50 POP_TOP
             52 JUMP_ABSOLUTE           12
        >>   54 LOAD_CONST               0 (None)
             56 RETURN_VALUE
```

</details>

But I was more interested in the readability of it and think this is better. (Although what probably made it harder to understand for me was the long variable name - `lang_code_split` - I always find that the longer it is, the less clear it is.)

Anyway, even if you still reject this, I just thought it was a nice use of `:=` and the `stop` argument of `str.rfind()`. Python has many gems like this that are overlooked. 😄

Reply:
@pope1ni IMO, it's an elegant solution and a nice use of the walrus operator.
If walrus operator is a problem (I like it in this case, but it's a controversial feature, as it breaks backward comparability, and the resulted code seems less 'pythonic' and more C-like, which might be harder to read for some people), you can rewrite it as:
```suggestion
        i = len(lang_code)
        while i > -1:
            i = lang_code.rfind('-', 0, i)
            possible_lang_codes.append(lang_code[:i])
```
Even in this state, it's better than what I originally wrote, because it doesn't create a new list and doesn't use `join`, which should be better for readability (no long variable names), uses less memory and should, in theory, perform better.

Reply:
@Cup0fCoffee Thanks. Your version is also good if we're rejecting mine because it has the walrus operator.

> If walrus operator is a problem (I like it in this case, but it's a controversial feature, as it breaks backward compatibility, and the resulted code seems less 'pythonic' and more C-like, which might be harder to read for some people)

Ah, yes. I never understand this argument of things being "unpythonic" because a feature may have come from another language. Python, like many languages, is evolving all the time and it makes sense to take good ideas from other places. The same negative argument is playing out for addition of pattern matching. Typically those making these claims are used to the language as it is and don't want to see it change or have another thing to learn. Often they are teaching beginners and think these changes are too complex to teach while disregarding those who are experienced and have a need for these features. At the end of the day people can use which language features they like and ignore those they don't if they're anti.

I would love to see [PEP 505](https://www.python.org/dev/peps/pep-0505/) come to fruition. Granted it is more syntactical sugar, but it allows for writing much more concise code in some cases. The argument against is that it doesn't read as clearly to some, but then many layers of nested blocks are less clear too, at least in my opinion. Readability is also subjective, a matter of perception - something that is readable to "you" might not be as readable to me.

In the case of `:=`, the token is uniquely used for assignment expressions and cannot be confused with `=` as would be possible in C. The backward compatibility bit is moot as we're Python 3.8+ only on master. I can't see how else it would be backward incompatible.

(Please don't take this as me implying you're for or against in any way... Was just stating my viewpoint on the subject.)

Reply:
Thanks both :+1: I pushed edits.